### PR TITLE
Breaking change check: New request required default param on existing path

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
This check will notify a breaking change if a new required default param was added to a request on an existing modified path.

### CODE REVIEW

**Comment by reuvenharrison** [comment]
Hi @Jonathan-Rosenberg and thanks for the nice contribution.
I haven't reviewed the code yet but I noticed that the new check isn't being run by the oasdiff cmd-line.
To make it run you need to add it to the defaultChecks function:
https://github.com/jonathan-rosenberg/oasdiff/blob/e3b94fcb8f4b9a163072fd49dcc0262c2956e3ec/checker/default_checks.go#L61

**Comment by Jonathan-Rosenberg** [comment]
Hi @reuvenharrison,
Forgot to `git add` it. 
Thanks!

File: checker/check-new-request-non-path-default-parameter.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	}
+	for path, pathItem := range diffReport.PathsDiff.Modified {
+		if pathItem.ParametersDiff == nil {
+			continue
+		}
+		for paramLoc, paramNameList := range pathItem.ParametersDiff.Added {
+			if paramLoc == "path" {
+				continue
+			}
+			for _, param := range pathItem.Revision.Parameters {
```
Comment (reuvenharrison) [inline checker/check-new-request-non-path-default-parameter.go:22]: Please check that Revision isn't nil before using it

Comment (Jonathan-Rosenberg) [reply checker/check-new-request-non-path-default-parameter.go:22]: Done

**Comment by reuvenharrison** [comment]
It seems like this test produces a false-positive when there are no endpoints under the path.
For example, comparing the two specs below reports a breaking-change error which in fact isn't breaking since there are no endpoints.
While this is a rare case, I still think we should take it into considration.
```
info:
  title: Tufin
  version: 1.0.0
openapi: 3.0.3
paths:
  /api/test:
    parameters:
      - in: header
        name: param1
        required: false
```

To:
```
info:
  title: Tufin
  version: 1.0.0
openapi: 3.0.3
paths:
  /api/test:
    parameters:
      - in: header
        name: param2
        required: true
```

**Comment by Jonathan-Rosenberg** [comment]
Hi @reuvenharrison, 
Added a check and a test for the edge case.
Thanks

File: checker/check-new-request-non-path-default-parameter.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		if pathItem.ParametersDiff == nil || pathItem.Revision == nil || len(pathItem.Revision.Operations()) == 0 {
+			continue
+		}
+		for paramLoc, paramNameList := range pathItem.ParametersDiff.Added {
+			if paramLoc == "path" {
+				continue
+			}
+			for _, param := range pathItem.Revision.Parameters {
+				if !paramNameList.Contains(param.Value.Name) {
+					continue
```
Comment (reuvenharrison) [inline checker/check-new-request-non-path-default-parameter.go]: I suggest to report this error on each affected endpoint rather than once per path.
The reason is that ApiChange is designed to report a breaking change at the endpoint (operation level), for example, it has an operation field.
Another reason is that the breaking change may only apply to a subset of the operations, for exampple, consider moving a param from op level to path level.

The downside to this approach is that the error may be reported multiple times, one for each operation, but this is ok since we will eventually support the ability to group breaking changes (see https://github.com/Tufin/oasdiff/issues/371).

File: checker/check-new-request-non-path-default-parameter.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+				}
+				id := "new-required-request-default-parameter-to-existing-path"
```
Comment (reuvenharrison) [inline checker/check-new-request-non-path-default-parameter.go:24]: This still need to be implemented.

Comment (reuvenharrison) [reply checker/check-new-request-non-path-default-parameter.go:24]: I'll add a new issue to handle this

**Review by reuvenharrison — APPROVED** [review]
This check refers to a path rather than an endpoint and because of that the resulting ApiChange lacks the Operation and OperationID fields.
Moreover,

File: checker/check-new-request-non-path-default-parameter.go [hunk 4]
```diff
@@ -13,26 +13,35 @@ func NewRequestNonPathDefaultParameterCheck(diffReport *diff.Diff, _ *diff.Opera
...
+		for paramLoc := range pathItem.ParametersDiff.Added {
 			if paramLoc == "path" {
 				continue
 			}
+
 			for _, param := range pathItem.Revision.Parameters {
-				if !paramNameList.Contains(param.Value.Name) {
-					continue
-				}
+
```
Comment (Jonathan-Rosenberg) [inline checker/check-new-request-non-path-default-parameter.go]: The removal of this code makes the check incorrect.
The level will be `ERR` even if the parameter wasn't added.
I'm creating a fix.

Comment (reuvenharrison) [reply checker/check-new-request-non-path-default-parameter.go]: I was under the impression that it is redundant because we are iterating on pathItem.ParametersDiff.Added
Can you replicate the issue with a unit test?

Comment (Jonathan-Rosenberg) [reply checker/check-new-request-non-path-default-parameter.go]: We do that, but since the `pathItem.Revision.Parameters` field is a slice and not a map, we traverse through it to find the `ParameterRef` object that matches the name we found in the `Added` map.
[PR](https://github.com/Tufin/oasdiff/pull/384).
